### ISSUE TITLE
Implement Dependency RFCs 0003-0006: Dependency Management

### ISSUE THREAD

**Issue by sophiewigmore** [paketo-buildpacks/rfcs#236]
This is the top-level tracking issue for implementing the Dependency Management RFC suite:
* [Overview RFC](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0003-dependency-management-overview.md)
* [Phase 1 RFC](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md)
* [Phase 2 RFC](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0005-dependency-management-phase-two.md)
* [Phase 3 RFC](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0006-dependency-management-phase-three.md)

## Phase 1 - Adopting Federated Model
A dependency-specific top-level tracking issue for each dependency in the list can be filed on the buildpack repository of interest using the following template.

<details>
<summary>Issue Template</summary>

#### Title: Implement Dependency RFC 0004 for \<Dependency\>
-------------------------------------------------------------

```
Implement [Dependency Management RFC Phase 1](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md) for \<INSERT DEPENDENCY\>. Check out the RFC for more details and background.

### 1. Determine dependency source strategy
<details>
<summary>Background</summary>

When possible, dependencies should be used directly from upstream, rather than undergoing any additional compilation or modifications performed by Paketo-maintained code. For each dependency, the corresponding buildpack maintainer group will decide if the dependency can be used directly from upstream, and must identify the location from which the dependency will be pulled from. Some of the Paketo Java buildpacks perform directory stripping during the buildpack build process itself. This could be a viable alternative to performing directory modifications during the dependency management process for maintainers to consider.
</details>

- [x] Determine whether the dependency can be used directly from its upstream, rather than undergoing additional compilation/modification. 
* If using directly from upstream, note where dependencies will come from
* If compiled or modified, note the decision for this
- [x] Document decision in a language-family level RFC. This decision can be combined in the RFC with the decision made for other buildpacks in the language family.

### 2. Version retrieval and metadata generation code
Please refer to the [retrieval RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#1-version-retrieval-make-retrieve) on version retrieval for details around the API and background.

- [x] Create retrieval code for the dependency
* Will live in the buildpack under: `dependency/retrieval/`
* Per [the RFC](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#transfercombine-version-retrieval-and-metadata-generation), this will be a combination of (1) discovering new dependency versions based on the `buildpack.toml`, and (2) generating metadata for each new version.
* **Note** : Per [the RFC caveat](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#caveat-compiled-dependencies) - if the dependency is to be compiled, the `SHA256` and `URI` field from the metadata should be omitted in this step.

### 3. Compilation code (if needed)
If the dependency will be compiled/modified, then refer to the [compilation RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#2-compilation) for API details and background. 

- [x] Create compilation action
* Will live in the buildpack under: `dependency/actions/compile/`

### 4. Dependency testing (optional)
It's up to maintainer discretion if the dependency will be tested. It's recommended to test the dependency if it's been compiled. Check out the [testing RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#3-test-dependency-make-test) for details.

- [ ] Add tests for dependency
* Will live in the buildpack under: `dependency/test.`

### 5. Makefile Setup
When using the generalized workflows for dependency management down the line, version retrieval and dependency testing will be executed via a Makefile in order to provide the workflow a standardized way to run the code, regardless of what language it was written in. Check out the [RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#overarching-plan) for what this should look like.

- [ ] Add Makefile
* Will live in the buildpack at: `dependency/Makefile`

### 6. Leveraging the new code
This issue serves to set up all the main logic for the dependency management. The work to actually leverage this code and migrate off of the dep-server will be completed in a separate issue once workflows and infrastructure is set up.
```
</details>

Dependency List:
- [x] bundler - https://github.com/paketo-buildpacks/bundler/issues/386 
- [X] composer - https://github.com/paketo-buildpacks/composer/issues/55 
- [X] curl - https://github.com/paketo-buildpacks/passenger/issues/254
- [ ] ~~dotnet-aspnetcore - https://github.com/paketo-buildpacks/dotnet-core-aspnet/issues/368~~
- [ ] ~~dotnet-runtime - https://github.com/paketo-buildpacks/dotnet-core-runtime/issues/372~~
- [X] dotnet-sdk - https://github.com/paketo-buildpacks/dotnet-core-sdk/issues/390
- [X] go - https://github.com/paketo-buildpacks/go-dist/issues/435
- [x] httpd - https://github.com/paketo-buildpacks/httpd/issues/355
- [X] icu - https://github.com/paketo-buildpacks/icu/issues/288
- [x] nginx - https://github.com/paketo-buildpacks/nginx/issues/414
- [X] node - https://github.com/paketo-buildpacks/node-engine/issues/508
- [X] php - https://github.com/paketo-buildpacks/php-dist/issues/415
- [X] pip - https://github.com/paketo-buildpacks/pip/issues/289
- [X] pipenv - https://github.com/paketo-buildpacks/pipenv/issues/252
- [x] python - https://github.com/paketo-buildpacks/cpython/issues/393
- [x] poetry - https://github.com/paketo-buildpacks/poetry/issues/86
- [X] ruby - https://github.com/paketo-buildpacks/mri/issues/393
- [x] rust - https://github.com/paketo-community/rust-dist/issues/199
- [x] tini - https://github.com/paketo-buildpacks/tini/issues/223
- [X] yarn - https://github.com/paketo-buildpacks/yarn/issues/261

## Phase 2 - Workflow and Github Action Generalization
- [x] Create new `jam` command for assembly - https://github.com/paketo-buildpacks/jam/issues/104
- [x] Create new dependency update workflow - https://github.com/paketo-buildpacks/github-config/issues/529

## Phase 3 - Migrating to the New System
- [x] (Per dependency) Disable the old `update-dependencies` workflow - once Phase 1 is complete for a dependency, and Phase 2 is complete, the old `update-dependencies` workflow can be disabled for the repository in the Github Actions UI.
- [x] Remove the old `update-dependencies` workflow from github-config - once every dependency has been moved over to the new system, remove the [old update-dependencies workflow](https://github.com/paketo-buildpacks/github-config/blob/dd6982d571546a26be6f1acf6b417ad34afc4e89/implementation/.github/workflows/update-dependencies.yml) all together.
- [ ] ~~Remove the old `jam update-dependencies` command - once every dependency has been moved over to the new system, the new `jam` command created in Phase 2 can fully replace the old command.~~
- [ ] Archive the dep-server repository - once every dependency in the dep-server has been moved to the new system, the dep-server repository can be archived so that the workflows stop running.

**Comment by sophiewigmore** [comment]
At this point, the dependency management system has been implemented. The final step of spinning down the dep-server and removing related tooling is now represented by separate issues and this issue can be closed out.

https://github.com/paketo-buildpacks/dep-server/issues/270
https://github.com/paketo-buildpacks/jam/issues/196
